### PR TITLE
Reintroduce accidentally deleted assertions.

### PR DESCRIPTION
**Pull request by oli-obk**
These were removed in https://github.com/rust-lang/rust/pull/50198

### CODE REVIEW

**Review by davidtwco — APPROVED** [review]
r=me when CI passes

**Comment by eddyb** [comment]
I would like to see a crater run on this, perhaps? Maybe there is some constant in the wild that we've let through despite our algorithms disagreeing on the exact bitpattern?

**Comment by oli-obk** [comment]
This is very small and easily reverted. I think we should just merge it and see what the next beta crater says

File: compiler/rustc_mir_build/src/thir/constant.rs [hunk 1]
```diff
@@ -61,20 +62,40 @@ fn parse_float<'tcx>(num: Symbol, fty: ty::FloatTy, neg: bool) -> Result<ConstVa
...
+            assert!(
+                u128::from(rust_f.to_bits()) == f.to_bits(),
+                "apfloat::ieee::Single gave different result for `{}`: \
+                 {}({:#x}) vs Rust's {}({:#x})",
+                rust_f,
+                f,
+                f.to_bits(),
+                Single::from_bits(rust_f.to_bits().into()),
+                rust_f.to_bits()
+            );
```
Comment (klensy) [inline compiler/rustc_mir_build/src/thir/constant.rs:78]: Is this multiple recalculating `rust_f.to_bits()`, `f.to_bits()` cheap in this place?

Comment (eddyb) [reply compiler/rustc_mir_build/src/thir/constant.rs:78]: `to_bits` is effectively free, but also this code isn't peformance-sensitive (compared to the cost of actually *parsing* the floating-point values in the first place).

**Comment by jackh726** [comment]
I feel like if this is something we *might* expect breakage on, we probably should rollup? Even if it's really small. (Then we can bisect to this exact change?)

**Pull request merged**